orc8r/cloud/go/services/ctraced/obsidian/handlers/handlers.go

```diff
@@ -0,0 +1,246 @@
+ */
+
+package handlers
+
+import (
+	"net/http"
+
+	"magma/orc8r/cloud/go/obsidian"
+	"magma/orc8r/cloud/go/orc8r"
+	"magma/orc8r/cloud/go/serdes"
+	"magma/orc8r/cloud/go/services/configurator"
+	"magma/orc8r/cloud/go/services/ctraced/obsidian/models"
+	"magma/orc8r/cloud/go/services/ctraced/storage"
+	merrors "magma/orc8r/lib/go/errors"
+	"magma/orc8r/lib/go/protos"
+
+	"github.com/labstack/echo"
+	"github.com/pkg/errors"
+)
+
+const (
+	Tracing = "tracing"
+	// v1/networks/:network_id/tracing
+	TracingRootPath = obsidian.V1Root + obsidian.MagmaNetworksUrlPart + obsidian.UrlSep + ":" + pathParamNetworkID + obsidian.UrlSep + Tracing
+	// v1/networks/:network_id/tracing/:trace_id
+	TracingPath = TracingRootPath + obsidian.UrlSep + ":" + pathParamTraceID
+	// v1/networks/:network_id/tracing/:trace_id/download
+	TracingDownloadPath = TracingPath + obsidian.UrlSep + "download"
+
+	pathParamTraceID   = "trace_id"
+	pathParamNetworkID = "network_id"
+)
+
+func GetObsidianHandlers(client GWCtracedClient, storage storage.CtracedStorage) []obsidian.Handler {
+	//client := NewGwCtracedClientImpl()
+	ret := []obsidian.Handler{
+		{Path: TracingRootPath, Methods: obsidian.GET, HandlerFunc: ListCallTraces},
+		{Path: TracingRootPath, Methods: obsidian.POST, HandlerFunc: GetCreateCallTraceHandlerFunc(client)},
+		{Path: TracingPath, Methods: obsidian.GET, HandlerFunc: GetCallTrace},
+		{Path: TracingPath, Methods: obsidian.PUT, HandlerFunc: GetUpdateCallTraceHandlerFunc(client, storage)},
+		{Path: TracingPath, Methods: obsidian.DELETE, HandlerFunc: GetDeleteCallTraceHandlerFunc(client, storage)},
+		{Path: TracingDownloadPath, Methods: obsidian.GET, HandlerFunc: GetDownloadCallTraceHandlerFunc(storage)},
+	}
+
+	return ret
+}
+
+func ListCallTraces(c echo.Context) error {
+	networkID, nerr := obsidian.GetNetworkId(c)
+	if nerr != nil {
+		return nerr
+	}
+
+	callTraces, err := configurator.LoadAllEntitiesOfType(
+		networkID, orc8r.CallTraceEntityType,
+		configurator.EntityLoadCriteria{LoadConfig: true},
+		serdes.Entity,
+	)
+	if err != nil {
+		return obsidian.HttpError(err, http.StatusInternalServerError)
+	}
+
+	ret := map[string]*models.CallTrace{}
+	for _, ctEnt := range callTraces {
+		ret[ctEnt.Key] = (&models.CallTrace{}).FromEntity(ctEnt)
+	}
+	return c.JSON(http.StatusOK, ret)
+}
+
+func GetCreateCallTraceHandlerFunc(client GWCtracedClient) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		networkID, nerr := obsidian.GetNetworkId(c)
+		if nerr != nil {
+			return nerr
+		}
+		cfg := &models.CallTraceConfig{}
+		if err := c.Bind(cfg); err != nil {
+			return obsidian.HttpError(err, http.StatusBadRequest)
+		}
+		ctr := &models.CallTrace{
+			Config: cfg,
+			State: &models.CallTraceState{
+				CallTraceAvailable: false,
+				CallTraceEnding:    false,
+			},
+		}
+		if err := ctr.ValidateModel(); err != nil {
+			return obsidian.HttpError(err, http.StatusBadRequest)
+		}
+
+		req, err := buildStartTraceRequest(cfg)
```

Comment:
Should we check if a call trace ID is already existing ?

Reply:
Sounds good